### PR TITLE
feat(person): add notes to person

### PR DESCRIPTION
**Pull request by jnizet**
Everything is there now, and can be reviewed.

### CODE REVIEW

**Comment by acrepet** [comment]
I've tested : LGTM !

**Comment by jnizet** [comment]
@cexbrayat I added tests fo the frontend code, and lost a huge time on the tests of the person-notes component, especially the test that checks that a note is updated. I added a comment in the code, in the beforeEach, explaining what I had to do to make that work, out of frustration. But if you have an idea on why it doesn't work by just spying on the update method of the service in the test, and can make it work that way, I would love to have the explanation.

**Comment by jnizet** [comment]
Other interesting fact: fakeAsync tests don't work when using Observable.delay(). Lost quite some time on that too.

File: frontend/src/app/person-notes/person-notes.component.spec.ts [hunk 1]
```diff
@@ -1,7 +1,294 @@
...
+
+  it('should display notes', () => {
+    const personNoteService = TestBed.get(PersonNoteService);
+    personNoteService.list.and.returnValue(Observable.of(notes));
+
+    const fixture = TestBed.createComponent(PersonNotesComponent);
+    fixture.componentInstance.person = person;
+    fixture.detectChanges();
+
+    expect(fixture.nativeElement.querySelectorAll('gl-note').length).toBe(2);
```
Comment (cexbrayat) [inline frontend/src/app/person-notes/person-notes.component.spec.ts]: `expect(fixture.debugElement.queryAll(By.directive(NoteComponent)).length).toBe(2)` ?

Comment (cexbrayat) [inline frontend/src/app/person-notes/person-notes.component.spec.ts]: Same below ?

File: frontend/src/app/person-notes/person-notes.component.spec.ts [hunk 2]
```diff
@@ -1,7 +1,294 @@
...
+    personNoteService.list.and.returnValue(Observable.of(notes));
+    personNoteService.delete.and.returnValue(Observable.of(null));
+    spyOn(confirmService, 'confirm').and.returnValue(Observable.throw('nok'));
+
+    const fixture = TestBed.createComponent(PersonNotesComponent);
+    fixture.componentInstance.person = person;
+    fixture.detectChanges();
+
+    // delete first note and don't confirm
+    const noteComponents = fixture.debugElement.queryAll(By.directive(NoteComponent));
```
Comment (cexbrayat) [inline frontend/src/app/person-notes/person-notes.component.spec.ts]: If you only need the first one, you can just `query`

File: frontend/src/app/person-notes/person-notes.component.ts [hunk 3]
```diff
@@ -31,10 +32,11 @@ export class PersonNotesComponent implements OnInit {
               private confirmService: ConfirmService) { }
 
   ngOnInit(): void {
-    const notes$ = this.personNoteService.list(this.person.id).share();
-    // display the spinner after 300ms, unless the notes have loaded before
-    Observable.of(true).delay(300).takeUntil(notes$).subscribe(() => this.spinnerDisplayed = true);
-    notes$.subscribe(notes => {
+    // display the spinner after 300ms, unless the notes have loaded before. Note: Observable.delay is untestable,
+    // see https://github.com/angular/angular/issues/10127
```
Comment (cexbrayat) [inline frontend/src/app/person-notes/person-notes.component.ts]: Yup we had the same issue in the ebook and Pro Pack see https://github.com/Ninja-Squad/Formation/issues/217

File: frontend/src/app/person-notes/person-notes.component.spec.ts [hunk 4]
```diff
@@ -1,7 +1,294 @@
...
+  let notes: Array<NoteModel>;
+
+  beforeEach(async(() => {
+    TestBed.configureTestingModule({
+      imports: [FormsModule, HttpClientModule, NgbModule.forRoot()],
+      declarations: [PersonNotesComponent, NoteComponent],
+      providers: [
+        // For a reason that I can't understand, using the real implementation and spying on its update
+        // method doesn't work. The real method is being called.
+        { provide: PersonNoteService, useValue: createSpyObj('PersonService', ['list', 'create', 'update', 'delete']) },
```
Comment (cexbrayat) [inline frontend/src/app/person-notes/person-notes.component.spec.ts]: I gave it a try real quick, and I don't have a problem if I just provide `PsersonNoteService`, and then replace the spies by `spyOn(personNoteService, 'list').and...`. The tests are OK.

Comment (jnizet) [inline frontend/src/app/person-notes/person-notes.component.spec.ts]: Actually, they're not, even now. I realized, looking at the code coverage, that saveNote() is never being called. The closure passed to whenStable().then() is never executed.

**Comment by jnizet** [comment]
I need to get back to the tests calling saveNote. I really don't understand what's happening there, but I'll try figuring it out. Cedric, if you have an idea, don't hesitate.

**Comment by cexbrayat** [comment]
@jnizet wrap in `async` ?

**Comment by jnizet** [comment]
Yep, that's what missing, and there are a few bugs in the test, too. I'll commit soon.

**Pull request merged**